clang/lib/CConv/ConstraintBuilder.cpp

```diff
@@ -344,6 +357,72 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
     }
   }
 
+  // Get the type variable used in a parameter declaration, or return null if no
+  // type variable is used.
+  const TypeVariableType *getTypeVariableType(ParmVarDecl *ParmDecl){
+    // This makes a lot of assumptions about how the AST will look.
+    if(InteropTypeExpr *PTy = ParmDecl->getInteropTypeExpr())
+      if (const clang::Type *T = PTy->getType().getTypePtr())
+        if (const auto *PtrTy =
+            dyn_cast_or_null<TypedefType>(T->getPointeeType().getTypePtr()))
+          return dyn_cast<TypeVariableType>(PtrTy->desugar());
+    return nullptr;
+  }
+
+  // Collect the set of TypeVariableTypes that are always used for arguments
+  // with the same type. These are type variables that can be instantiated with
+  // a concrete type, so it is correct to remove casts to void* on their
+  // arguments.
+  // TODO: Check that the type parameter for the return type agrees with its
+  //       other uses. This is non-trivial because the type of the CallExpr is
+  //       always void* for itype generics. I'll need to add something in
+  //       vistCastExpr to remember how itype generics are used. When making
+  //       this change, look out for ways to avoid duplicating code from
+  //       AllocTypeParamAdder in RewriteUtils.
+  void getInstantiableTypeParams(CallExpr *CE,
+                                 FunctionDecl *FD,
+                                 std::set<const TypeVariableType *> &Types) {
+    assert("Must provide nonnull FunctionDecl." && FD);
+    // Construct a map from TypeVariables to the set of arguments that must have
+    // that type.
+    std::map<const TypeVariableType *,
+             std::set<const clang::Type *>> TypeVarClasses;
+    unsigned int I = 0;
+    for (auto *const A : CE->arguments()) {
+      // This can happen with varargs
+      if(I >= FD->getNumParams())
+        break;
+      if (const auto *Ty = getTypeVariableType(FD->getParamDecl(I)))
+        TypeVarClasses[Ty].insert(ignoreCast(A)->getType().getTypePtr());
```

Comment:
I'm confused why it's possible for a type variable to have multiple types. If you have a function call and the argument parameter is type T, which you have quantified, then T can only be instantiated to one thing. You can't have `forall T. T*T -> T` but then instantiate `T` to both `int *` and `float *` for the same call. What is going on here?

Reply:
OK, I see what's going on. You are confirming that a type variable is being instantiated consistently. Only in that case can the cast be removed.

I would rename the function to something like `getConsistentlyInstantiatedTypeParams`. This is because you are looking at a particular call, and thus particularly instantiations, and you are making sure they are consistent.